### PR TITLE
Fix Travisci build

### LINKED ISSUE #36: Build is broken
[Build is broken](https://travis-ci.org/edmundmok/mealpy). This was introduced when I added [`check-requirements`](https://github.com/edmundmok/mealpy/blob/b1f27a46681e3072094c811d1655ca34b4690c06/Makefile#L17) for #13 .

The issue is that [ipython](https://github.com/edmundmok/mealpy/blob/b1f27a46681e3072094c811d1655ca34b4690c06/requirements-dev-minimal.txt#L3) installs conditional dependencies based on OS platform and `requirements-tools` doesn't handle this edge case. Options for fixing:
* Move `check-requirements` from `make test`, maybe make it the responsibility of dev who updates requirements to run. It's only a sanity/safety check 
* Remove ipython from requirements-dev. That and pudb are optional tools that should be installed adhoc, per developer setup. Maybe restrict requirements-dev to only tooling *necessary* to deploy, i.e. `pytest`

### PR DESCRIPTION
**Pull request by ipwnponies**
The build broke because it's running on xenial for python 3.6 and 3.7. Python packages can have conditional dependencies which are platform-dependent, which makes it impossible to run `check-requirements` using the same `requirements-minimal`. i.e. `ipython` [installs a host of conditional dpenedencies](https://github.com/ipython/ipython/blob/f14271fe2b0f9627f64848466de944befc7c29c9/setup.py#L202-L206).

My proposed fix is:
* remove non-essential packages from `requirements-dev.txt`, only what is necessary to run test or do stuff. Devs can elect to install packages on their own
* Target xenial python 3.7 for `check-requirements`. This is arbitrary and probably non-issue 98% of the time. But occasionally the build will break and we'll need to remember to appease xenial py37 overlord

Finally, I added a `make venv-dev`, which I'm fine with removing. i'm wondering if it would be useful to have a single command to install "all optional" tooling. But i understand this can be contentious and opinionated in what is allowed in there. It's a separate commit that can be dropped.

Fixes #36

### CODE REVIEW

**Review by edmundmok — COMMENTED** [review]
> remove non-essential packages from requirements-dev.txt, only what is necessary to run test or do stuff. Devs can elect to install packages on their own

Definitely for this, otherwise the dev requirements can end up very bloated.

> Target xenial python 3.7 for check-requirements. This is arbitrary and probably non-issue 98% of the time. But occasionally the build will break and we'll need to remember to appease xenial py37 overlord

Does this mean if someone wants to update `requirements-dev` from `requirements-dev-minimal` they have to be on py3.7?

> Finally, I added a make venv-dev, which I'm fine with removing. i'm wondering if it would be useful to have a single command to install "all optional" tooling. But i understand this can be contentious and opinionated in what is allowed in there. It's a separate commit that can be dropped.

This should be okay, may be useful for some and won't hurt for others who may not use it.

**Comment by ipwnponies** [comment]
> > Target xenial python 3.7 for check-requirements. This is arbitrary and probably non-issue 98% of the time. But occasionally the build will break and we'll need to remember to appease xenial py37 overlord
> 
> Does this mean if someone wants to update `requirements-dev` from `requirements-dev-minimal` they have to be on py3.7?

I would recommend they do it in py3.7 but it's very likely py3.6 will still generate the same output. Most packages don't have platform-specific dependencies. i.e. if we need datetime parsing library, it shouldn't need different supporting packages for different OSes.

Some packages install backport dependencies. For example, python 3 has `functools.lru_cache` in stdlib but python 2 requires a backport package. If you run `upgrade-requirements` in python 2, it will include `functools32` in output but not in python 3. The difference between python 3.6 and 3.7 is small (I can only think of  `dataclasses`).

Long way of justifying the 98%. But I acknowledge it's possible this will cause issues, I plan to keep an eye on changes to requirements.

**Comment by edmundmok** [comment]
> I would recommend they do it in py3.7 but it's very likely py3.6 will still generate the same output. Most packages don't have platform-specific dependencies. i.e. if we need datetime parsing library, it shouldn't need different supporting packages for different OSes.

In that case, it shouldn't be a problem to test check-requirements against 3.6 too right? The issue was caused by non-essential packages, and the essential ones, at least currently, do not have these conditional dependencies.

**Comment by ipwnponies** [comment]
I haven't verified if the remaining packages on 3.6 have issue. Even if `check-requirements` works for 3.6 now, it's a matter of time before it diverges and we'll be right back here and need to make a call.

We could run `check-requirements` for both but it's not that valuable. `check-requirements` is a sanity check that we've pinned everything that will be installed, to avoid issues where user reports bug but it's caused because of slightly different install setup than the one we've developed for. I don't want to make it hard requirement, since we don't control where and how it's going to be installed (no containerization).

This is more of a code quality check, so that future PRs will have automated mechanism of enforcing "did you fully pin all dependencies". If working around it becomes an issue, I'm in favour of disabling it and just running it adhoc from time-to-time instead.

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ install:
   - make venv
 script:
   - make test
+  - test $TRAVIS_PYTHON_VERSION == '3.7' && venv/bin/check-requirements
```
Comment (edmundmok) [inline .travis.yml]: This is [failing](https://travis-ci.org/edmundmok/mealpy/builds/527588437) on python 3.6:
```
$ test $TRAVIS_PYTHON_VERSION == '3.7' && venv/bin/check-requirements
The command "test $TRAVIS_PYTHON_VERSION == '3.7' && venv/bin/check-requirements" exited with 1.
```
Maybe `test $TRAVIS_PYTHON_VERSION != '3.7' || venv/bin/check-requirements`?

**Review by edmundmok — COMMENTED** [review]
Btw, if they do diverge and someone using python 3.6 requires a different version of some package than what is listed in `requirements-dev`, would that cause any issues? If it does cause issues then it would be good to have the CI detect that. I've not seen any issues like that before usually since the level of specificity I use is what is in `requirements-dev-minimal` and just let pip handle any further dependencies, whereas the generated `requirements-dev` is very explicit in all the dependencies used. 

Actually, I assume pip can probably handle that scenario though since currently different packages may already depend on different versions of the same library and it still works fine. I think running check-requirements just on 3.7 should be okay then.

File: .travis.yml (hunk 1 above)
Comment (ipwnponies) [reply .travis.yml]: This is just me being bad at shell scripting. I want `if test $TRAVIS_PYTHON_VERSION == '3.7'; venv/bin/check-requirements; fi`. In 3.6, it'll noop as expected, instead of failing on `test`.

**Comment by ipwnponies** [comment]
> Btw, if they do diverge and someone using python 3.6 requires a different version of some package than what is listed in `requirements-dev`, would that cause any issues? 

Likely won't cause issues, as we're talking about **extra** dependencies being added and for platform-specific needs. The ones we do know about are pinned, for reproducible environments.

> ... I use is what is in `requirements-dev-minimal` and just let pip handle any further dependencies, whereas the generated `requirements-dev` is very explicit in all the dependencies used.

What you had before is 0% pinned dependencies. `check-requirements` is for 100% pinned. My stance, from the previous comments, is to try to aim for 100% but be okay with 90-ish% coverage. Pin the majority of dependencies but if a couple extra are installed as additional platform-specific dependencies, I'm fine with ignoring them.
 
> Actually, I assume pip can probably handle that scenario though since currently different packages may already depend on different versions of the same library and it still works fine.

Oh boy, you're in for a surprise, pip doesn't do [dependency management](https://pip.pypa.io/en/stable/user_guide/#requirements-files) at all:
>... pip doesn’t have true dependency resolution, but instead simply uses the first specification it finds for a project

It's why we're using `venv-update` + `requirements-tools`, which will error out if conflicting dependencies are installed/generated.
